speechbrain/lobes/models/synthesis/tacotron2/model.py

```diff
@@ -0,0 +1,1184 @@
+        )
+        self.memory_layer = LinearNorm(
+            embedding_dim, attention_dim, bias=False, w_init_gain="tanh"
+        )
+        self.v = LinearNorm(attention_dim, 1, bias=False)
+        self.location_layer = LocationLayer(
+            attention_location_n_filters,
+            attention_location_kernel_size,
+            attention_dim,
+        )
+        self.score_mask_value = -float("inf")
+
+    def get_alignment_energies(
+        self, query, processed_memory, attention_weights_cat
+    ):
+        """
+        Arguments
+        ---------
+        query: torch.Tensor
+            decoder output (batch, n_mel_channels * n_frames_per_step)
+        processed_memory: torch.Tensor
+            processed encoder outputs (B, T_in, attention_dim)
+        attention_weights_cat: torch.Tensor
+            cumulative and prev. att weights (B, 2, max_time)
+
+        Returns
+        -------
+        alignment : torch.Tensor
+            (batch, max_time)
+        """
+
+        processed_query = self.query_layer(query.unsqueeze(1))
+        processed_attention_weights = self.location_layer(attention_weights_cat)
+        energies = self.v(
+            torch.tanh(
+                processed_query + processed_attention_weights + processed_memory
+            )
+        )
+
+        energies = energies.squeeze(2)
+        return energies
+
+    def forward(
+        self,
+        attention_hidden_state,
+        memory,
+        processed_memory,
+        attention_weights_cat,
+        mask,
+    ):
+        """
+        Arguments
+        ---------
+        attention_hidden_state: torch.Tensor
+            attention rnn last output
+        memory: torch.Tensor
+            encoder outputs
+        processed_memory: torch.Tensor
+            processed encoder outputs
+        attention_weights_cat: torch.Tensor
+            previous and cummulative attention weights
+        mask: torch.Tensor
+            binary mask for padded data
+        """
+        alignment = self.get_alignment_energies(
+            attention_hidden_state, processed_memory, attention_weights_cat
+        )
+
+        alignment = alignment.masked_fill(mask, self.score_mask_value)
+
+        attention_weights = F.softmax(alignment, dim=1)
+        attention_context = torch.bmm(attention_weights.unsqueeze(1), memory)
+        attention_context = attention_context.squeeze(1)
+
+        return attention_context, attention_weights
+
+
+class Prenet(nn.Module):
+    def __init__(self, in_dim, sizes):
+        super(Prenet, self).__init__()
+        in_sizes = [in_dim] + sizes[:-1]
+        self.layers = nn.ModuleList(
+            [
+                LinearNorm(in_size, out_size, bias=False)
+                for (in_size, out_size) in zip(in_sizes, sizes)
+            ]
+        )
+
+    def forward(self, x):
+        for linear in self.layers:
+            x = F.dropout(F.relu(linear(x)), p=0.5, training=True)
+        return x
+
+
+class Postnet(nn.Module):

```

Comment:
@flexthink  

- Can we make the code more modular by separating blocks like PostNet Block, Input Text Pre-processing part in different files at common place , so it can be used for FastSpeech as well.

Reply:
Definitely. I believe this one comes from the original NVIDIA Tacotron.

We already have some code sharing across models
For example,

* `speechbrain/lobes/models/synthesis/dataio.py`: shared
* `speechbrain/lobes/models/synthesis/<model>/dataio.py`: specific

So far, we haven't tried sharing NN blocks, but we could create

* `speechbrain/lobes/models/synthesis/model.py`

for anything that's shared

we could even split `model.py` into multiple files for a single model - and for the shared ones too if the files get too big.